### DISCUSSION TITLE
jax.debug.impure_callback

### DISCUSSION THREAD

**Discussion by pchanial** [jax-ml/jax#22278] [Ideas]
In order to polish some informative messages and because jax.debug.callback is traced, I resorted to jax.experimental.io_callback to print a conditional message:
```python
def callback(solution):
    niter = solution['niter']
    if niter < nitermax:
        print(f'Converged in {niter} iterations')
    else:
        print(f'Did not converge in {niter} iteration')
    return jax.array(0, dtype=np.int8)
```
which is then called in a staged out function with:
```python
_ = jax.experimental.io_callback(callback, jax.ShapeDtypeStruct((), np.int8), solution)
```
It would be cleaner to just have to write it without the trick of returning a useless scalar with a function `jax.debug.impure_callback(callback, solution)`

**Comment by jakevdp** [comment]
It sounds like what you're asking for already exists in `jax.debug.callback`. Can you say more about why this isn't sufficient for your use-case?

**Reply by pchanial** [reply]
My apologies, I cannot reproduce the problem I encountered. I will reopen the discussion if I see again the issue.
